### DISCUSSION TITLE
[BUG] GPU DETECTED 0

### DISCUSSION THREAD

**Discussion by BoTaliu** [autogluon/autogluon#4600] [Q&A]
when I add `num_gpus` as follows and run the code

`predictor = TabularPredictor(label=label_column,log_to_file=True,verbosity=2)
predictor.fit(train_data=data, presets='best_quality',num_gpus=1)`

it shows
![image](https://github.com/user-attachments/assets/1460e734-94f4-4b1d-b4c6-a6fab82764a2)

I use windows 10, python --version == 3.11.10 and autogluon --version == 1.1.1, GPU is 4060ti, CUDA version == 12.6. And I test `torch.cuda.is_available()`, it returns **True**. And I am wondering whether it is because I don't have cuda.dll as follows in my computer
![image](https://github.com/user-attachments/assets/0486cb9e-fb7f-4966-a43d-e14ce548f026)

**Comment by celestinoxp** [comment]
I have similar issue, i see: "Specified total num_gpus: 1, but only 0 are available. Will use 0 instead", 

![image](https://github.com/user-attachments/assets/c400b4ad-1a51-43f7-9b55-e345c4762bc5)

but when use a code to check if gpu is available it show a gpu (with cuda).

![image](https://github.com/user-attachments/assets/32aadc41-e6cb-43ea-91ea-c69c81dd2b0b)

i'm using a nightly latest code of autogluon.tabular:

![image](https://github.com/user-attachments/assets/1d1c29f9-5ae8-46aa-b70c-6ff868a527a6)

Anyone can check what's wrong?
@Innixma @LennartPurucker @tonyhoo

**Comment by LennartPurucker** [comment]
Heyho, thanks for reporting this! 

Can you tell me what the following commands return for you:

```python
import ray
print(ray.cluster_resources().get("GPU", None)) # Our GPU check for Ray
```

Also, as a fallback, going from the memory warning in your log, can you try passing `ag_args_ensemble={"fold_fitting_strategy": "sequential_local"}` to fit() to run AutoGluon without ray.

**Reply by celestinoxp** [reply]
Your suggested code show me:

---------------------------------------------------------------------------
RaySystemError                            Traceback (most recent call last)
Cell In[1], line 2
      1 import ray
----> 2 print(ray.cluster_resources().get("GPU", None)) # Our GPU check for Ray

File [~\anaconda3\Lib\site-packages\ray\_private\client_mode_hook.py:103](http://localhost:8888/lab/tree/totoloto/principal/~/anaconda3/Lib/site-packages/ray/_private/client_mode_hook.py#line=102), in client_mode_hook.<locals>.wrapper(*args, **kwargs)
    101     if func.__name__ != "init" or is_client_mode_enabled_by_default:
    102         return getattr(ray, func.__name__)(*args, **kwargs)
--> 103 return func(*args, **kwargs)

File [~\anaconda3\Lib\site-packages\ray\_private\state.py:994](http://localhost:8888/lab/tree/totoloto/principal/~/anaconda3/Lib/site-packages/ray/_private/state.py#line=993), in cluster_resources()
    982 @DeveloperAPI
    983 @client_mode_hook
    984 def cluster_resources():
    985     """Get the current total cluster resources.
    986 
    987     Note that this information can grow stale as nodes are added to or removed
   (...)
    992             resource in the cluster.
    993     """
--> 994     return state.cluster_resources()

File [~\anaconda3\Lib\site-packages\ray\_private\state.py:732](http://localhost:8888/lab/tree/totoloto/principal/~/anaconda3/Lib/site-packages/ray/_private/state.py#line=731), in GlobalState.cluster_resources(self)
    722 def cluster_resources(self):
    723     """Get the current total cluster resources.
    724 
    725     Note that this information can grow stale as nodes are added to or
   (...)
    730             resource in the cluster.
    731     """
--> 732     self._check_connected()
    734     # Calculate total resources.
    735     total_resources = defaultdict(int)

File [~\anaconda3\Lib\site-packages\ray\_private\state.py:53](http://localhost:8888/lab/tree/totoloto/principal/~/anaconda3/Lib/site-packages/ray/_private/state.py#line=52), in GlobalState._check_connected(self)
     51 # _really_init_global_state should have set self.global_state_accessor
     52 if self.global_state_accessor is None:
---> 53     raise ray.exceptions.RaySystemError(
     54         "Ray has not been started yet. You can start Ray with 'ray.init()'."
     55     )

RaySystemError: System error: Ray has not been started yet. You can start Ray with 'ray.init()'.

**Reply by celestinoxp** [reply]
after that i changed your sugested code, inserting a new line with "ray.init()", and it works now. You can see:
![image](https://github.com/user-attachments/assets/0b4f24c0-7805-4625-82a9-81199accc3b6)

**Reply by LennartPurucker** [reply]
Mhm, that's interesting. So, the log message you showed above uses this call. Hence, AutoGluon should know you have a GPU if this call works. 

Did `ag_args_ensemble={"fold_fitting_strategy": "sequential_local"}` work for you? 

Another point: it looks like the log message is printed when trying to fit the final ensemble. This model does not use GPUs anyway, so this should not affect AutoGluon's performance. Can you share a full log message with `verbosity=4` at fit()?

**Reply by celestinoxp** [reply]
Did ag_args_ensemble={"fold_fitting_strategy": "sequential_local"} work for you? --> i think work normal with all cpu cores

today i updated autogluon again: first uninstall with (pip uninstall autogluon.tabular) and installed latest version using: "uv pip install --pre autogluon.tabular -U"

i found a strange use of integrated gpu, as you can see:
![image](https://github.com/user-attachments/assets/083bfbcb-076c-4d4f-8ad2-ba0213b23551)

i want use use the "gpu 0" (nvidia, because have cuda) instead "gpu 1" (amd, integrated gpu)... 

now i will send verbosity=4 details, so i user normal code:
![image](https://github.com/user-attachments/assets/849f83ad-f701-4bed-b74f-04565be27c7f)

and complete log (i used time_limit=1000 for this report, normally i ise 4000 ):

[log-verbosity4.txt](https://github.com/user-attachments/files/17659139/log-verbosity4.txt)

and if you need check packages versions:
i used:
from autogluon.core.utils import show_versions
show_versions()

and show:
INSTALLED VERSIONS
------------------
date                : 2024-11-07
time                : 09:21:25.226886
python              : 3.11.9.final.0
OS                  : Windows
OS-release          : 10
Version             : 10.0.22631
machine             : AMD64
processor           : AMD64 Family 23 Model 96 Stepping 1, AuthenticAMD
num_cores           : 16
cpu_ram_mb          : 15791.171875
cuda version        : None
num_gpus            : 1
gpu_ram_mb          : [3935]
avail_disk_size_mb  : None

autogluon           : None
autogluon.common    : 1.1.2b20241106
autogluon.core      : 1.1.2b20241106
autogluon.features  : 1.1.2b20241106
autogluon.tabular   : 1.1.2b20241106
boto3               : 1.35.55
catboost            : 1.2.7
fastai              : 2.7.18
hyperopt            : 0.2.7
imodels             : 2.0.0
lightgbm            : 4.5.0
matplotlib          : 3.10.0rc1
networkx            : 3.4.2
numpy               : 1.26.4
onnx                : 1.17.0
onnxruntime         : 1.20.0
onnxruntime-gpu     : 1.20.0
pandas              : 2.2.3
psutil              : 6.0.0
pyarrow             : 17.0.0
ray                 : 2.31.0
requests            : 2.32.3
scikit-learn        : 1.5.2
scikit-learn-intelex: 2025.0.0
scipy               : 1.12.0
setuptools          : 75.3.0
skl2onnx            : 1.17.0
spacy               : 3.7.5
tabpfn              : 0.1.10
torch               : 2.5.1+cu121
tqdm                : 4.67.0
vowpalwabbit        : None
xgboost             : 2.1.2

**Reply by celestinoxp** [reply]
![image](https://github.com/user-attachments/assets/45e0e916-a0e3-43c0-9862-9968dd434d84)

What means "Windows fatal exception: code 0xc0000139" ?

**Reply by celestinoxp** [reply]
Update:
i found a possible expanation:

now i'm using: ag_args_ensemble={"fold_fitting_strategy": "parallel_local"}, and the expanation maybe autogluon changed automatically  so sequential_local that only uses cpu...

![image](https://github.com/user-attachments/assets/32fa47c1-3f3e-4ee7-a446-097ffab4a00b)

now with ag_args_ensemble={"fold_fitting_strategy": "parallel_local"} i see my gpu in use:

![image](https://github.com/user-attachments/assets/107670ea-dc34-42f3-bfa3-8e9633d8a994)

summary: to use the gpu we must use ag_args_ensemble={"fold_fitting_strategy": "parallel_local"}, to force to use the gpu

It seems like this is the solution, but I think the autogluon code should be analyzed, and add a message to the log explaining why the gpu is not used, according to the context I explained before.

**Reply by LennartPurucker** [reply]
So, from the log, I can see that AutoGluon is not using the GPU in several cases because it would be less efficient than using the CPUs.

If we only fit one model at the same time, we can also pass the GPU and use it. But if we have 8 CPUs and want to fit 8 models in parallel, it is better to ignore the GPU.

**Reply by LennartPurucker** [reply]
Adding a message/warning that we do not use GPUs (because it is not a good idea) is a good suggestion; thanks! I will note it down, but also feel free to open an issue/feature suggestion with it to have your name on it and track its progress.

**Reply by celestinoxp** [reply]
@LennartPurucker i think i found the main problem, https://github.com/autogluon/autogluon/issues/4623

**Reply by LennartPurucker** [reply]
@celestinoxp, well spotted; thank you for bringing this up and reporting on it; it was very helpful!

**Comment by celestinoxp** [comment]
@BoTaliu If you feel we are clear, you can close this discussion...

**Reply by BoTaliu** [reply]
Thanks, bro. I see your contribution and the discussions. It enlightens me a lot.
And I will close the discussion.
